e2e flake: starting test fails with panic: runtime error: index out of range [0] with length 0

I see this failure many times when running the tests in the first time with a new environment. Typically the next tries work. Looks like incorrect setup that fixes itself on the next run since a resource was created on the cluster in the failed attempt.

We can add a retry for the entire test suite, it is faster to rerun the tests than to create a new environment.

```
=== RUN   TestSuites/Exhaustive/Deploy-rbd#01/Appset
2024-09-16T22:16:15.056+0530	INFO	deployers/applicationset.go:17	enter Deploy appset-deploy-cephfs-busybox
=== PAUSE TestSuites/Exhaustive/Deploy-rbd#01/Appset
=== CONT  TestSuites/Exhaustive/Deploy-rbd#01/Appset
=== RUN   TestSuites/Exhaustive/Deploy-rbd#01/Appset/Deploy
=== RUN   TestSuites/Exhaustive/Deploy-rbd/Subscr
2024-09-16T22:16:15.056+0530	INFO	deployers/applicationset.go:17	enter Deploy appset-deploy-rbd-busybox
=== PAUSE TestSuites/Exhaustive/Deploy-rbd/Subscr
=== RUN   TestSuites/Exhaustive/Deploy-cephfs/Subscr/Deploy
2024-09-16T22:16:15.056+0530	INFO	deployers/subscription.go:31	enter Deploy subscr-deploy-cephfs-busybox
=== PAUSE TestSuites/Exhaustive/Deploy-cephfs#02/Disapp
=== CONT  TestSuites/Exhaustive/Deploy-cephfs#02/Disapp
=== RUN   TestSuites/Exhaustive/Deploy-rbd#02/Disapp
=== CONT  TestSuites/Exhaustive/Deploy-rbd/Subscr
=== PAUSE TestSuites/Exhaustive/Deploy-rbd#02/Disapp
=== CONT  TestSuites/Exhaustive/Deploy-rbd#02/Disapp
=== RUN   TestSuites/Exhaustive/Deploy-rbd#02/Disapp/Deploy
2024-09-16T22:16:15.0[57](https://github.com/RamenDR/ramen/actions/runs/10887936359/job/30211267322#step:12:58)+0530	INFO	deployers/discoveredapps.go:24	enter Deploy disapp-deploy-rbd-busybox
=== RUN   TestSuites/Exhaustive/Deploy-rbd/Subscr/Deploy
=== NAME  TestSuites/Exhaustive/Deploy-cephfs#02/Disapp
    exhaustive_suite_test.go:93: Workload Deploy-cephfs not supported by deployer Disapp, skip test
2024-09-16T22:16:15.057+0530	INFO	deployers/subscription.go:31	enter Deploy subscr-deploy-rbd-busybox
=== RUN   TestSuites/Exhaustive/Deploy-cephfs#01/Appset/Enable
2024-09-16T22:16:16.920+0530	INFO	dractions/actions.go:38	enter EnableProtection appset-deploy-cephfs-busybox
=== RUN   TestSuites/Exhaustive/Deploy-rbd#01/Appset/Enable
2024-09-16T22:16:16.936+0530	INFO	dractions/actions.go:38	enter EnableProtection appset-deploy-rbd-busybox
2024-09-16T22:16:17.022+0530	INFO	util/crud.go:99	channel ramen-gitops is deleted
--- FAIL: TestSuites (2.99s)
    --- PASS: TestSuites/Validate (0.38s)
        --- PASS: TestSuites/Validate/CheckRamenHubOperatorStatus (0.32s)
        --- PASS: TestSuites/Validate/CheckRamenSpokeOperatorStatus (0.06s)
    --- FAIL: TestSuites/Exhaustive (2.35s)
        --- PASS: TestSuites/Exhaustive/Deploy-cephfs#02 (0.00s)
            --- SKIP: TestSuites/Exhaustive/Deploy-cephfs#02/Disapp (0.00s)
        --- FAIL: TestSuites/Exhaustive/Deploy-rbd#01 (1.92s)
            --- FAIL: TestSuites/Exhaustive/Deploy-rbd#01/Appset (1.92s)
                --- PASS: TestSuites/Exhaustive/Deploy-rbd#01/Appset/Deploy (1.88s)
                --- FAIL: TestSuites/Exhaustive/Deploy-rbd#01/Appset/Enable (0.04s)
panic: runtime error: index out of range [0] with length 0 [recovered]
	panic: runtime error: index out of range [0] with length 0
```

--- a/e2e/dractions/actions.go
+++ b/e2e/dractions/actions.go
@@ -43,18 +43,13 @@ func EnableProtection(w workloads.Workload, d deployers.Deployer) error {
 	placementName := name
 	drpcName := name
 
-	placementDecisionName, err := waitPlacementDecision(util.Ctx.Hub.CtrlClient, namespace, placementName)
-	if err != nil {
-		return err
-	}
-
-	placementDecision, err := getPlacementDecision(util.Ctx.Hub.CtrlClient, namespace, placementDecisionName)
+	placementDecision, err := waitPlacementDecision(util.Ctx.Hub.CtrlClient, namespace, placementName)
 	if err != nil {
 		return err
 	}
 
 	clusterName := placementDecision.Status.Decisions[0].ClusterName
-	util.Ctx.Log.Info("got clusterName " + clusterName + " from " + placementDecisionName)
+	util.Ctx.Log.Info("got clusterName " + clusterName + " from " + placementDecision.Name)
 
 	util.Ctx.Log.Info("update placement " + placementName + " annotation")
 

--- a/e2e/dractions/crud.go
+++ b/e2e/dractions/crud.go
@@ -33,18 +33,6 @@ func updatePlacement(client client.Client, placement *clusterv1beta1.Placement) 
 	return client.Update(context.Background(), placement)
 }
 
-func getPlacementDecision(client client.Client, namespace, name string) (*clusterv1beta1.PlacementDecision, error) {
-	placementDecision := &clusterv1beta1.PlacementDecision{}
-	key := types.NamespacedName{Namespace: namespace, Name: name}
-
-	err := client.Get(context.Background(), key, placementDecision)
-	if err != nil {
-		return nil, err
-	}
-
-	return placementDecision, nil
-}
-
 func getDRPC(client client.Client, namespace, name string) (*ramen.DRPlacementControl, error) {
 	drpc := &ramen.DRPlacementControl{}
 	key := types.NamespacedName{Namespace: namespace, Name: name}

--- a/e2e/dractions/retry.go
+++ b/e2e/dractions/retry.go
@@ -15,38 +15,28 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-// nolint:gocognit
-// return placementDecisionName, error
+// waitPlacementDecision waits until we have a placement decision and returns the placement decision object.
 func waitPlacementDecision(client client.Client, namespace string, placementName string,
-) (string, error) {
+) (*v1beta1.PlacementDecision, error) {
 	startTime := time.Now()
-	placementDecisionName := ""
 
 	for {
 		placement, err := getPlacement(client, namespace, placementName)
 		if err != nil {
-			return "", err
+			return nil, err
 		}
 
-		for _, cond := range placement.Status.Conditions {
-			if cond.Type == "PlacementSatisfied" && cond.Status == "True" {
-				placementDecisionName = placement.Status.DecisionGroups[0].Decisions[0]
-				if placementDecisionName != "" {
-					return placementDecisionName, nil
-				}
-			}
-		}
-
-		// if placement is controlled by ramen, it will not have placementdecision name in its status
-		// so need query placementdecision by label
 		placementDecision, err := getPlacementDecisionFromPlacement(client, placement)
-		if err == nil && placementDecision != nil {
-			return placementDecision.Name, nil
+		if err != nil {
+			return nil, err
+		}
+
+		if placementDecision != nil && len(placementDecision.Status.Decisions) > 0 {
+			return placementDecision, nil
 		}
 
 		if time.Since(startTime) > util.Timeout {
-			return "", fmt.Errorf(
-				"could not get placement decision for " + placementName + " before timeout, fail")
+			return nil, fmt.Errorf("timeout waiting for placement decisions for %q ", placementName)
 		}
 
 		time.Sleep(util.RetryInterval)
@@ -135,19 +125,12 @@ func waitDRPCPhase(client client.Client, namespace, name string, phase ramen.DRS
 }
 
 func getCurrentCluster(client client.Client, namespace string, placementName string) (string, error) {
-	placementDecisionName, err := waitPlacementDecision(client, namespace, placementName)
+	placementDecision, err := waitPlacementDecision(client, namespace, placementName)
 	if err != nil {
 		return "", err
 	}
 
-	placementDecision, err := getPlacementDecision(client, namespace, placementDecisionName)
-	if err != nil {
-		return "", err
-	}
-
-	clusterName := placementDecision.Status.Decisions[0].ClusterName
-
-	return clusterName, nil
+	return placementDecision.Status.Decisions[0].ClusterName, nil
 }
 
 // return dr cluster client

